src/qref/_schema_v1.py

```diff
@@ -96,6 +103,28 @@ class RoutineV1(BaseModel):
     def __init__(self, **data: Any):
         super().__init__(**{k: v for k, v in data.items() if v != [] and v != {}})
 
+    @field_validator("connections", mode="after")
+    @classmethod
+    def _validate_connections(cls, v, values) -> list[_ConnectionV1]:
+        for connection in v:
+            _validate_connection_end(connection.source, values)
+            _validate_connection_end(connection.target, values)
+        return v
```

Comment:
**issue**: Every iteration of the loop runs through all the children. This shouldn't be neccessary and this function could be simplified to:

```python
def _validate_connections(cls, v, values) -> list[_ConnectionV1]:
    children_port_names = [
        f"{child.name}.{port.name}"
        for child in values.data.get("children")
        for port in child.ports
     ]
     parent_port_names = [port.name for port in values.data["ports"]]
     available_port_names = set(children_port_names + parent_port_names)
    
     missed_ports = [
        port
        for connection in v
        for port in (connection.source, connection.target)
        if port not in available_port_names
    ]
    if missed_ports:
        raise ValueError(
            "The following ports appear in a connection but are not "
            "among routine's port or their children's ports: {missed_ports}."
        )

Reply:
Ok, implemented it!